// inngest/components/connect/strategies/sameThread/index.cjs
const require_connect = require('../../../../proto/src/components/connect/protobuf/connect.cjs');
const require_buffer = require('../../buffer.cjs');
const require_types = require('../../types.cjs');
const require_BaseStrategy = require('../core/BaseStrategy.cjs');
const require_connection = require('../core/connection.cjs');

//#region src/components/connect/strategies/sameThread/index.ts
const ResponseAcknowledgeDeadline = 5e3;
/**
* Same-thread connection strategy.
*
* This strategy runs the WebSocket connection, heartbeater, and lease extender
* in the same thread as user code execution. This is the default strategy.
*/
var SameThreadStrategy = class extends require_BaseStrategy.BaseStrategy {
	config;
	messageBuffer;
	core;
	constructor(config) {
		super({ logger: config.internalLogger });
		this.config = config;
		this.core = new require_connection.ConnectionCore({
			apiBaseUrl: config.apiBaseUrl,
			appIds: Object.keys(config.requestHandlers),
			connectionData: config.connectionData,
			envName: config.envName,
			hashedFallbackKey: config.hashedFallbackKey,
			hashedSigningKey: config.hashedSigningKey,
			instanceId: config.options.instanceId,
			maxWorkerConcurrency: config.options.maxWorkerConcurrency,
			mode: config.mode,
			gatewayUrl: config.options.gatewayUrl
		}, {
			logger: this.internalLogger,
			onStateChange: (state) => {
				if (this._state === require_types.ConnectionState.CLOSING || this._state === require_types.ConnectionState.CLOSED) return;
				this._state = state;
			},
			getState: () => this._state,
			handleExecutionRequest: async (request) => {
				const handler = this.config.requestHandlers[request.appName];
				if (!handler) throw new Error(`No handler for app: ${request.appName}`);
				const response = await handler(request);
				const responseBytes = require_connect.SDKResponse.encode(response).finish();
				this.messageBuffer.addPending(request.requestId, responseBytes, ResponseAcknowledgeDeadline);
				return responseBytes;
			},
			onReplyAck: (requestId) => {
				this.messageBuffer.acknowledgePending(requestId);
			},
			onBufferResponse: (requestId, responseBytes) => {
				this.messageBuffer.append(requestId, responseBytes);
			},
			onConnectionActive: (signingKey) => {
				this.messageBuffer.flush(signingKey);
			}
		});
		this.messageBuffer = new require_buffer.MessageBuffer({
			envName: config.envName,
			getApiBaseUrl: () => this.core.getApiBaseUrl(),
			logger: this.internalLogger
		});
	}
	get connectionId() {
		return this.core.connectionId;
	}
	getDebugState() {
		return this.core.getDebugState();
	}
	async connect(attempt = 0) {
		this.throwIfClosingOrClosed();
		this.setupShutdownSignalIfConfigured(this.config.options.handleShutdownSignals);
		try {
			await this.messageBuffer.flush(this.config.hashedSigningKey);
		} catch (err) {
			this.internalLogger.debug({ err }, "Failed to flush messages, using fallback key");
			await this.messageBuffer.flush(this.config.hashedFallbackKey);
		}
		await this.core.start(attempt);
	}
	async close() {
		this.cleanupShutdown();
		this.setClosing();
		await this.core.close();
		this.internalLogger.debug("Flushing messages before closing");
		try {
			await this.messageBuffer.flush(this.config.hashedSigningKey);
		} catch (err) {
			this.internalLogger.debug({ err }, "Failed to flush messages, using fallback key");
			await this.messageBuffer.flush(this.config.hashedFallbackKey);
		}
		this.setClosed();
		this.internalLogger.debug("Fully closed");
	}
};

//#endregion
exports.SameThreadStrategy = SameThreadStrategy;
