// inngest/components/connect/index.cjs
const require_consts = require('../../helpers/consts.cjs');
const require_env = require('../../helpers/env.cjs');
const require_Inngest = require('../Inngest.cjs');
const require_config = require('./config.cjs');
const require_types = require('./types.cjs');
const require_index = require('./strategies/index.cjs');

//#region src/components/connect/index.ts
/**
* WebSocket worker connection that implements the WorkerConnection interface.
*
* This class acts as a facade that delegates to a connection strategy.
* The strategy determines how the WebSocket connection, heartbeater, and
* lease extender are managed (same thread vs worker thread).
*/
var WebSocketWorkerConnection = class {
	inngest;
	options;
	strategy;
	constructor(options) {
		if (!Array.isArray(options.apps) || options.apps.length === 0 || !options.apps[0]) throw new Error("No apps provided");
		this.inngest = options.apps[0].client;
		for (const app of options.apps) {
			const client = app.client;
			if (client.env !== this.inngest.env) throw new Error(`All apps must be configured to the same environment. ${client.id} is configured to ${client.env} but ${this.inngest.id} is configured to ${this.inngest.env}`);
		}
		this.options = this.applyDefaults(options);
	}
	applyDefaults(opts) {
		const options = { ...opts };
		if (!Array.isArray(options.handleShutdownSignals)) options.handleShutdownSignals = require_types.DEFAULT_SHUTDOWN_SIGNALS;
		const env = require_env.getProcessEnv();
		if (options.maxWorkerConcurrency === void 0) {
			const envValue = env[require_consts.envKeys.InngestConnectMaxWorkerConcurrency];
			if (envValue) {
				const parsed = Number.parseInt(envValue, 10);
				if (!Number.isNaN(parsed) && parsed > 0) options.maxWorkerConcurrency = parsed;
			}
		}
		if (options.isolateExecution === void 0) {
			const envValue = env[require_consts.envKeys.InngestConnectIsolateExecution];
			if (envValue === "0" || envValue === "false") options.isolateExecution = false;
		}
		if (options.gatewayUrl === void 0) {
			const envValue = env[require_consts.envKeys.InngestConnectGatewayUrl];
			if (envValue) options.gatewayUrl = envValue;
		}
		return options;
	}
	get state() {
		return this.strategy?.state ?? require_types.ConnectionState.CONNECTING;
	}
	get connectionId() {
		if (!this.strategy?.connectionId) throw new Error("Connection not prepared");
		return this.strategy.connectionId;
	}
	get closed() {
		if (!this.strategy) throw new Error("No connection established");
		return this.strategy.closed;
	}
	getDebugState() {
		if (!this.strategy) return {
			state: require_types.ConnectionState.CONNECTING,
			activeConnectionId: void 0,
			drainingConnectionId: void 0,
			lastHeartbeatSentAt: void 0,
			lastHeartbeatReceivedAt: void 0,
			lastMessageReceivedAt: void 0,
			shutdownRequested: false,
			inFlightRequestCount: 0,
			inFlightRequests: []
		};
		return this.strategy.getDebugState();
	}
	async close() {
		if (!this.strategy) return;
		return this.strategy.close();
	}
	/**
	* Establish a persistent connection to the gateway.
	*/
	async connect(attempt = 0) {
		this.inngest[require_Inngest.internalLoggerSymbol].debug({ attempt }, "Establishing connection");
		const { hashedSigningKey, hashedFallbackKey, envName, connectionData, requestHandlers } = require_config.prepareConnectionConfig(this.options.apps, this.inngest);
		this.strategy = await require_index.createStrategy({
			hashedSigningKey,
			hashedFallbackKey,
			internalLogger: this.inngest[require_Inngest.internalLoggerSymbol],
			envName,
			connectionData,
			requestHandlers,
			options: this.options,
			apiBaseUrl: this.inngest.apiBaseUrl,
			mode: this.inngest["mode"]
		}, this.options);
		await this.strategy.connect(attempt);
	}
};
const connect = async (options) => {
	if (options.apps.length === 0) throw new Error("No apps provided");
	const conn = new WebSocketWorkerConnection(options);
	await conn.connect();
	return conn;
};

//#endregion
exports.connect = connect;
