// inngest/components/connect/strategies/core/heartbeat.cjs
const require_connect = require('../../../../proto/src/components/connect/protobuf/connect.cjs');
const require_buffer = require('../../buffer.cjs');
const require_types = require('./types.cjs');

//#region src/components/connect/strategies/core/heartbeat.ts
var HeartbeatManager = class {
	interval;
	intervalMs = 1e4;
	onHeartbeatSent;
	constructor(accessor, wakeSignal, logger) {
		this.accessor = accessor;
		this.wakeSignal = wakeSignal;
		this.logger = logger;
	}
	/**
	* Update the heartbeat interval. Restarts the timer if the interval changed
	* or if it wasn't running yet.
	*/
	updateInterval(ms) {
		if (ms === this.intervalMs && this.interval) return;
		this.intervalMs = ms;
		this.stop();
		this.start();
	}
	/** Stop the heartbeat timer. */
	stop() {
		clearInterval(this.interval);
		this.interval = void 0;
	}
	start() {
		if (this.interval) return;
		this.interval = setInterval(() => this.tick(), this.intervalMs);
	}
	tick() {
		const conn = this.accessor.activeConnection;
		if (!conn || conn.ws.readyState !== WebSocket.OPEN) return;
		if (conn.pendingHeartbeats >= 2) {
			this.logger.warn({ connectionId: conn.id }, "Consecutive heartbeats missed, reconnecting");
			conn.dead = true;
			this.wakeSignal.wake(require_types.WAKE_REASON.HeartbeatMissed);
			return;
		}
		conn.pendingHeartbeats++;
		conn.ws.send(require_buffer.ensureUnsharedArrayBuffer(require_connect.ConnectMessage.encode(require_connect.ConnectMessage.create({ kind: require_connect.GatewayMessageType.WORKER_HEARTBEAT })).finish()));
		this.logger.debug({ connectionId: conn.id }, "Heartbeat sent");
		this.onHeartbeatSent?.();
	}
};

//#endregion
exports.HeartbeatManager = HeartbeatManager;
