// inngest/components/execution/otel/consts.js
//#region src/components/execution/otel/consts.ts
const debugPrefix = "inngest:otel";
let TraceStateKey = /* @__PURE__ */ function(TraceStateKey$1) {
	TraceStateKey$1["AppId"] = "inngest@app";
	TraceStateKey$1["FunctionId"] = "inngest@fn";
	TraceStateKey$1["TraceRef"] = "inngest@traceref";
	return TraceStateKey$1;
}({});
let Attribute = /* @__PURE__ */ function(Attribute$1) {
	Attribute$1["InngestTraceparent"] = "inngest.traceparent";
	Attribute$1["InngestTraceRef"] = "inngest.traceref";
	Attribute$1["InngestRunId"] = "sdk.run.id";
	Attribute$1["InngestAppId1"] = "sdk.app.id";
	Attribute$1["InngestAppId2"] = "sys.app.id";
	Attribute$1["InngestFunctionId"] = "sys.function.id";
	Attribute$1["InngestStepParentSpanId"] = "inngest.step.parentSpanId";
	Attribute$1["InngestStepId"] = "inngest.step.id";
	Attribute$1["InngestStepIndex"] = "inngest.step.index";
	Attribute$1["InngestStepHash"] = "inngest.step.hash";
	Attribute$1["InngestStepAttempt"] = "inngest.step.attempt";
	return Attribute$1;
}({});

//#endregion
export { Attribute, TraceStateKey, debugPrefix };
