// inngest/components/execution/otel/middleware.js
import { version } from "../../../version.js";
import { Middleware } from "../../middleware/middleware.js";
import { clientProcessorMap } from "./access.js";
import { debugPrefix } from "./consts.js";
import { createProvider, extendProvider } from "./util.js";
import Debug from "debug";
import { DiagLogLevel, diag, trace } from "@opentelemetry/api";

//#region src/components/execution/otel/middleware.ts
const devDebug = Debug(`${debugPrefix}:middleware`);
var InngestTracesLogger = class {
	#logger = Debug(`${debugPrefix}:diag`);
	debug = this.#logger;
	error = this.#logger;
	info = this.#logger;
	verbose = this.#logger;
	warn = this.#logger;
};
/**
* Middleware the captures and exports spans relevant to Inngest runs using
* OTel.
*
* This can be used to attach additional spans and data to the existing traces
* in your Inngest dashboard (or Dev Server).
*/
const extendedTracesMiddleware = ({ behaviour = "auto", instrumentations, logLevel = DiagLogLevel.ERROR } = {}) => {
	devDebug("behaviour:", behaviour);
	let processor;
	let processorReady;
	switch (behaviour) {
		case "auto": {
			const extended = extendProvider(behaviour);
			if (extended.success) {
				devDebug("extended existing provider");
				processor = extended.processor;
				break;
			}
			processorReady = createProvider(behaviour, instrumentations).then((created) => {
				if (created.success) {
					devDebug("created new provider");
					processor = created.processor;
				} else console.warn("no provider found to extend and unable to create one", created.error ?? "");
			});
			break;
		}
		case "createProvider":
			processorReady = createProvider(behaviour, instrumentations).then((created) => {
				if (created.success) {
					devDebug("created new provider");
					processor = created.processor;
				} else console.warn("unable to create provider, Extended Traces middleware will not work", created.error ?? "");
			});
			break;
		case "extendProvider": {
			const extended = extendProvider(behaviour);
			if (extended.success) {
				devDebug("extended existing provider");
				processor = extended.processor;
				break;
			}
			console.warn("unable to extend provider, Extended Traces middleware will not work. Either allow the middleware to create a provider by setting `behaviour: \"createProvider\"` or `behaviour: \"auto\"`, or make sure that the provider is created and imported before the middleware is used.");
			break;
		}
		case "off": break;
		default: console.warn(`unknown behaviour ${JSON.stringify(behaviour)}, defaulting to "off"`);
	}
	class ExtendedTracesMiddleware extends Middleware.BaseMiddleware {
		id = "inngest:extended-traces";
		/**
		* Called by the Inngest constructor to associate the processor with the
		* client.
		*/
		static onRegister({ client }) {
			devDebug("set otel diagLogger:", diag.setLogger(new InngestTracesLogger(), logLevel));
			if (processor) clientProcessorMap.set(client, processor);
			else if (processorReady) processorReady.then(() => {
				if (processor) clientProcessorMap.set(client, processor);
			}).catch((err) => {
				devDebug("failed to register processor for client:", err);
			});
		}
		transformFunctionInput(arg) {
			return {
				...arg,
				ctx: {
					...arg.ctx,
					tracer: trace.getTracer("inngest", version)
				}
			};
		}
		async wrapRequest({ next }) {
			return next().finally(() => processor?.forceFlush());
		}
	}
	return ExtendedTracesMiddleware;
};

//#endregion
export { extendedTracesMiddleware };
