// @opentelemetry/instrumentation-oracledb/build/src/constants.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 * Copyright (c) 2025, Oracle and/or its affiliates.
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanNames = void 0;
// Contains span names produced by instrumentation
// It lists the RPC names (suffix with _MSG like EXECUTE_MSG) and
// exported oracledb functions (like EXECUTE).
// These constants need to be in sync with what is generated by the
// 'oracledb' module.
var SpanNames;
(function (SpanNames) {
    SpanNames["CONNECT"] = "oracledb.getConnection";
    SpanNames["POOL_CONNECT"] = "oracledb.Pool.getConnection";
    SpanNames["POOL_CREATE"] = "oracledb.createPool";
    SpanNames["CONNECT_PROTOCOL_NEG"] = "oracledb.ProtocolMessage";
    SpanNames["CONNECT_DATATYPE_NEG"] = "oracledb.DataTypeMessage";
    SpanNames["CONNECT_AUTH_MSG"] = "oracledb.AuthMessage";
    SpanNames["CONNECT_FAST_AUTH"] = "oracledb.FastAuthMessage";
    SpanNames["EXECUTE_MSG"] = "oracledb.ExecuteMessage";
    SpanNames["EXECUTE"] = "oracledb.Connection.execute";
    SpanNames["EXECUTE_MANY"] = "oracledb.Connection.executeMany";
    SpanNames["LOGOFF_MSG"] = "oracledb.LogOffMessage";
    SpanNames["CONNECT_CLOSE"] = "oracledb.Connection.close";
    SpanNames["CREATE_LOB"] = "oracledb.Connection.createLob";
    SpanNames["LOB_MESSAGE"] = "oracledb.LobOpMessage";
    SpanNames["LOB_GETDATA"] = "oracledb.Lob.getData";
})(SpanNames = exports.SpanNames || (exports.SpanNames = {}));
